### PR TITLE
fix: unroll normalised ProgrammeMembershipDTOs

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
The assumption about the ProgrammeMembershipDTO containing only a single
CurriculumMembership was false.

TIS21-3076

### CODE REVIEW

File: src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java [hunk 1]
```diff
@@ -109,17 +114,20 @@ public List<DmsDto> assembleDmsDtos(List<Object> dtos) {
...
    *
    * @param dto The dto which will be mapped to another dto representative of the "data" portion of
    *            a DmsDto (e.g. PostDmsDto).
-   * @return The DmsDto, complete with data and metadata.
+   * @return The list of DmsDtos, complete with data and metadata.
    */
-  private DmsDto assembleDmsDto(Object dto) {
+  private List<DmsDto> assembleDmsDto(Object dto) {
+    List<DmsDto> dmsDtos = null;
+    boolean hasMultipleDmsDtos = false;
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java:130]: Not a thing of beauty :/

**Comment by ReubenRobertsHEE** [comment]
Marked as draft until https://github.com/Health-Education-England/TIS-SYNC/pull/176 is merged

File: src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java [hunk 2]
```diff
@@ -56,18 +67,55 @@ private DmsDto assembleDmsDto(Object dto) {
...
+    if (hasMultipleDmsDtos) {
+      return dmsDtos;
+    } else {
+      if (dmsData != null) {
+        MetadataDto metadata = new MetadataDto(Instant.now().toString(), DATA, LOAD,
+            PARTITION_KEY_TYPE, schema, table, UUID.randomUUID().toString());
 
-      return new DmsDto(dmsData, metadata);
+        return Collections.singletonList(new DmsDto(dmsData, metadata));
+      }
```
Comment (Judge40) [inline src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java:88]: A rethink of the DmsMapper interface might help here.

If we changed `O objectToDmsDto(Object object)` to `List<O> objectToDmsDtos(Object object)` we could expand the DTOs inside the mapper, then create the `DmsDto` including metadata for each item in the list.

Comment (Judge40) [reply src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java:88]: The mapper would be more complex but it feels an appropriate place to do it.
The alternative to keep the single DTO mappers simple could be to add a `multipleDto` flag to the enum and call `objectToDmsDto` or `objectToDmsDtos` based on that.

Comment (ReubenRobertsHEE) [reply src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java:88]: Let me take a look, that sounds like a good idea.

Comment (Judge40) [reply src/main/java/uk/nhs/tis/sync/service/DmsRecordAssembler.java]: That's pretty much it yeah, added a small comment where an update mapper could tidy up a few lines.

**Review by Judge40 — COMMENTED** [review]
LGTM

**Review by hihilary — APPROVED** [review]

**Pull request merged**